cmd/e2e-runner/main.go

```diff
@@ -174,6 +169,56 @@ func realMain(ctx context.Context) error {
 	return srv.ServeHTTPHandler(ctx, mux)
 }
 
+func bootstrap(ctx context.Context, db *database.Database, cfg *config.E2ERunnerConfig) mux.MiddlewareFunc {
+	logger := logging.FromContext(ctx).Named("e2erunner.bootstrap")
+	bootstrapFn := func() error {
+		mu.Lock()
+		defer mu.Unlock()
+
+		if !setupComplete {
+			// Bootstrap the environment
+			resp, err := envstest.Bootstrap(ctx, db)
+			if err != nil {
+				return fmt.Errorf("failed to bootstrap testsuite: %w", err)
+			}
+			// set the outer close function for later
+			closeFunc = func() error { return resp.Cleanup() }
+
+			// Verify that SMS is configured on the realm
+			if !project.SkipE2ESMS {
+				has, err := resp.Realm.HasSMSConfig(db)
+				if err != nil {
+					return fmt.Errorf("failed to check if realm has sms config: %w", err)
+				}
+				if !has {
+					return fmt.Errorf("realm does not have sms config, configure it or set E2E_SKIP_SMS to continue")
+				}
+			}
+
+			cfg.VerificationAdminAPIKey = resp.AdminAPIKey
```

Comment:
This is shared state, but it's only guarded by the mutex in bootstrap. The handlers also access and in some cases modify this resource. It might be fine, but I'm not 100% familiar with how the middleware is handled when there's shared state with a child handler? 

If we're going to do this, we might be better storing the config on the request context using context.WithValue inside the middleware?

I'm worried this is going to cause a subtle race condition that will be impossible to debug.

Reply:
The handlers don't get to run unless this has successfully run.

I could put more guards around it, but it's not strictly necessary. We already live with our config structs being pointers to shared state across multiple controllers and handlers.